File: src/platform/macos/window.zig
const objc = @import("objc");
const seam = @import("../backend.zig");

pub const Rect = extern struct { x: f64, y: f64, w: f64, h: f64 };
pub const Size = extern struct { w: f64, h: f64 };

/// Create an NSWindow with the given size. Returns the window id.
pub fn create(opts: seam.WindowOpts) objc.id {
    const NSWindow = objc.class("NSWindow");
    const win = objc.msgSend(*const fn (objc.Class, objc.SEL) callconv(.c) objc.id)(NSWindow, objc.sel("alloc"));
    const frame = Rect{ .x = 100, .y = 100, .w = opts.width, .h = opts.height };
    const styleMask: u64 = 1 | 2 | 4 | 8; // titled, closable, miniaturizable, resizable
    const initWin = objc.msgSend(*const fn (objc.id, objc.SEL, Rect, u64, i64, bool) callconv(.c) objc.id);
    const window = initWin(win, objc.sel("initWithContentRect:styleMask:backing:defer:"), frame, styleMask, 2, false);
    _ = objc.msgSend(*const fn (objc.id, objc.SEL, objc.id) callconv(.c) void)(window, objc.sel("setTitle:"), objc.nsString(opts.title.ptr));
    return window;
}

pub fn setTitle(window: objc.id, title: [:0]const u8) void {
    _ = objc.msgSend(*const fn (objc.id, objc.SEL, objc.id) callconv(.c) void)(window, objc.sel("setTitle:"), objc.nsString(title.ptr));
}

/// setContentSize: takes an NSSize ONLY (no display flag) (L7).
pub fn setSize(window: objc.id, w: f64, h: f64) void {
    _ = objc.msgSend(*const fn (objc.id, objc.SEL, Size) callconv(.c) void)(window, objc.sel("setContentSize:"), .{ .w = w, .h = h });
}

pub fn setFullscreen(window: objc.id, _: bool) void {
    _ = objc.msgSend(*const fn (objc.id, objc.SEL, objc.id) callconv(.c) void)(window, objc.sel("toggleFullScreen:"), null);
}

pub fn show(window: objc.id) void {
    _ = objc.msgSend(*const fn (objc.id, objc.SEL, bool) callconv(.c) void)(window, objc.sel("makeKeyAndOrderFront:"), true);
}

pub fn focus(window: objc.id) void {
    _ = objc.msgSend(*const fn (objc.id, objc.SEL, bool) callconv(.c) void)(window, objc.sel("makeKeyAndOrderFront:"), true);
}

File: frontend/window.js
(function () {
  const Z = window.Zigware || (window.Zigware = {});
  const listeners = new Map();
  Z._emit = function (event, payload) {
    const set = listeners.get(event);
    if (set) for (const fn of set) { try { fn(payload); } catch (e) {} }
  };
  function listen(event, cb) {
    let set = listeners.get(event);
    if (!set) listeners.set(event, (set = new Set()));
    set.add(cb);
    return function () { set.delete(cb); };
  }
  function makeWindow(label) {
    return {
      label,
      setTitle: (title) => Z.invoke("window.setTitle", { label, title }),
      setSize: (width, height) => Z.invoke("window.setSize", { label, width, height }),
      setFullscreen: (on) => Z.invoke("window.setFullscreen", { label, on }),
      focus: () => Z.invoke("window.focus", { label }),
      close: () => Z.invoke("window.close", { label }),
      listen,
    };
  }
  Z.Window = {
    getCurrent: () => makeWindow(window.__ZIGWARE_LABEL__ || "main"),
    getByLabel: (label) => makeWindow(label),
    create: (opts) => Z.invoke("window.create", opts).then(() => makeWindow(opts.label)),
    listenAll: (event, cb) => listen(event, cb),
  };
})();
